# ***Import Libraries, Save My API Key & Built My API URL***

In [ ]:
# Impoerted the requests libraries
# !pip install requests

import requests
from google.colab import userdata

# Saved my API key from Colab secrets
OPENWEATHER_API_KEY = userdata.get('OPENWEATHER_API_KEY')

# Built my base URL for the OpenWeatherMap API
BASE_URL = "http://api.openweathermap.org/data/2.5/weather"

# ***Extract Data***

I chose (3) cities- **London, NewYork & Tokyo**.
I extracted weather data in a structured Pandas DataFrame and ensure the column names are descriptive.

In [ ]:
import pandas as pd
import datetime

def get_weather_data(city_name):
    params = {
        'q': city_name,
        'appid': OPENWEATHER_API_KEY,
        'units': 'metric'  # You can change to 'imperial' for Fahrenheit
    }
    response = requests.get(BASE_URL, params=params)
    response.raise_for_status() # Raise an exception for HTTP errors
    return response.json()

cities = ["London", "New York", "Tokyo"]
weather_reports = []

for city in cities:
    try:
        weather_data = get_weather_data(city)

        # Extract relevant fields
        city_name = weather_data['name']
        temperature = weather_data['main']['temp']
        humidity = weather_data['main']['humidity']
        weather_condition = weather_data['weather'][0]['description'].capitalize()
        wind_speed = weather_data['wind']['speed']
        # Convert Unix timestamp to human-readable date and time
        dt_object = datetime.datetime.fromtimestamp(weather_data['dt'])
        date_time = dt_object.strftime('%Y-%m-%d %H:%M:%S')

        weather_reports.append({
            'City Name': city_name,
            'Temperature (°C)': temperature,
            'Humidity (%)': humidity,
            'Weather Condition': weather_condition,
            'Wind Speed (m/s)': wind_speed,
            'Date and Time': date_time
        })
    except requests.exceptions.RequestException as e:
        print(f"Error fetching data for {city}: {e}")
    except KeyError as e:
        print(f"Error parsing data for {city}: Missing key {e}")

# Create a pandas DataFrame from the collected data
df_weather = pd.DataFrame(weather_reports)

# Display the DataFrame
display(df_weather)

,City Name,Temperature (°C),Humidity (%),Weather Condition,Wind Speed (m/s),Date and Time
0,London,25.72,49,Overcast clouds,0.89,2026-07-22 13:51:24
1,New York,25.43,78,Scattered clouds,7.15,2026-07-22 13:50:24
2,Tokyo,31.14,78,Few clouds,2.05,2026-07-22 13:48:56


# ***Transform Data***

I transformed the data and ensure all data types are appropriate for analysis. Specifically, the 'Date and Time' column, currently a string, I converted them to a datetime object.

In [ ]:
print("Original Data Types:")
display(df_weather.info())

# Convert 'Date and Time' column to datetime objects
df_weather['Date and Time'] = pd.to_datetime(df_weather['Date and Time'])

print("\nData Types After Transformation:")
display(df_weather.info())

# Display the first few rows of the transformed DataFrame to confirm
print("\nTransformed DataFrame (first 5 rows):")
display(df_weather.head())

Original Data Types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   City Name          3 non-null      object 
 1   Temperature (°C)   3 non-null      float64
 2   Humidity (%)       3 non-null      int64  
 3   Weather Condition  3 non-null      object 
 4   Wind Speed (m/s)   3 non-null      float64
 5   Date and Time      3 non-null      object 
dtypes: float64(2), int64(1), object(3)
memory usage: 276.0+ bytes


None


Data Types After Transformation:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   City Name          3 non-null      object        
 1   Temperature (°C)   3 non-null      float64       
 2   Humidity (%)       3 non-null      int64         
 3   Weather Condition  3 non-null      object        
 4   Wind Speed (m/s)   3 non-null      float64       
 5   Date and Time      3 non-null      datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(2)
memory usage: 276.0+ bytes


None


Transformed DataFrame (first 5 rows):


,City Name,Temperature (°C),Humidity (%),Weather Condition,Wind Speed (m/s),Date and Time
0,London,25.79,49,Overcast clouds,0.89,2026-07-22 13:41:43
1,New York,25.11,80,Scattered clouds,5.36,2026-07-22 13:34:45
2,Tokyo,30.98,81,Few clouds,0.45,2026-07-22 13:43:16


# ***Load Data***

I saved the cleaned and transformed weather data for future analysis. I save it as a CSV file, which is a widely used.

In [ ]:
# Define the filename for the CSV file
csv_filename = 'weather_data.csv'

# Save the DataFrame to a CSV file
df_weather.to_csv(csv_filename, index=False)

print(f"Weather data successfully saved to {csv_filename}")

Weather data successfully saved to weather_data.csv


# ***Basic Analysis***

I proceeded in performing some basic analysis.

In [ ]:
import pandas as pd

# Perform Basic Analysis and consolidate findings for readability

print("## Basic Weather Analysis Summary\n")

# 1. Temperature Comparison
temp_comparison = df_weather[['City Name', 'Temperature (°C)']].sort_values(by='Temperature (°C)', ascending=False)
highest_temp_city = temp_comparison.iloc[0]
lowest_temp_city = temp_comparison.iloc[-1]

print("### 1. Temperature Analysis:")
print(f"- Highest Temperature: {highest_temp_city['Temperature (°C)']:.2f}°C in {highest_temp_city['City Name']}")
print(f"- Lowest Temperature: {lowest_temp_city['Temperature (°C)']:.2f}°C in {lowest_temp_city['City Name']}\n")


# 2. Humidity Comparison
humidity_comparison = df_weather[['City Name', 'Humidity (%)']].sort_values(by='Humidity (%)', ascending=False)
highest_humidity_city = humidity_comparison.iloc[0]

print("### 2. Humidity Analysis:")
print(f"- Highest Humidity: {highest_humidity_city['Humidity (%)']}% in {highest_humidity_city['City Name']}\n")


# 3. Weather Conditions Comparison
print("### 3. Weather Conditions Analysis:")
for index, row in df_weather.iterrows():
    print(f"- {row['City Name']}: {row['Weather Condition']}")

print("\n--- End of Analysis Summary ---")

## Basic Weather Analysis Summary

### 1. Temperature Analysis:
- Highest Temperature: 31.14°C in Tokyo
- Lowest Temperature: 25.43°C in New York

### 2. Humidity Analysis:
- Highest Humidity: 78% in New York

### 3. Weather Conditions Analysis:
- London: Overcast clouds
- New York: Scattered clouds
- Tokyo: Few clouds

--- End of Analysis Summary ---


# ***Tools Used***

**1. Python**: The primary programming language for data processing.

**2. Requests library**: Used for making HTTP requests to the OpenWeatherMap API.

**3. Pandas library**: Essential for data manipulation, structuring, and analysis (DataFrames).

**3. Datetime module**: For handling and converting date and time information.

**4. Google.colab.userdata**: For securely managing the API key.

**5. Openpyxl library (optional)**: For saving data to Excel format.

# ***Key Findings***
Based on the basic analysis of the collected weather data (at the time of extraction):

**1. Temperature Analysis**:
* Tokyo recorded the highest temperature at 31.14°C.
* New York recorded the lowest temperature at 24.43°C.

**2. Humidity**:
* New York had the highest humidity at 78%.

**3. Weather Conditions**:
* London: Overcast clouds.
* New York: Scattered clouds
* Tokyo: Few clouds.

